src/workerd/api/streams/writable-sink-adapter.h

```diff
@@ -0,0 +1,446 @@
+    BackpressureState(jsg::Promise<void>::Resolver&& resolver,
+        jsg::Promise<void>&& promise,
+        jsg::MemoizedIdentity<jsg::Promise<void>>&& watcher);
+
+    friend class WritableStreamSinkJsAdapter;
+    // Note that if the BackpressureState is dropped while in a waiting state,
+    // the ready promise will be left unresolved. This is OK.
+  };
+  BackpressureState backpressureState;
+  kj::Rc<WeakRef<WritableStreamSinkJsAdapter>> selfRef;
+
+  // Replaces the backpressure state with a new one, indicating that backpressure
+  // is being applied. If we are already in a backpressure state, this is a no-op.
+  // This will cause the ready promise (and its stable identity) to change.
+  void maybeSignalBackpressure(jsg::Lock& js);
+
+  // Conditionally releases backpressure if the desired size is now > 0.
+  void maybeReleaseBackpressure(jsg::Lock& js);
+
+  // Creates a new BackpressureState in the waiting state.
+  static BackpressureState newBackpressureState(jsg::Lock& js);
+};
+
+// ================================================================================
+
+// Adapts a WritableStream to a KJ-frendly interface.
+// The adapter fully wraps the WritableStream instance,
+// using a WritableStreamDefaultWriter to push data to it.
+// Then the adapter is destroyed or aborted, the writer is
+// aborted and both the writer and the stream references
+// are dropped. Critically, the stream is not usable after
+// ownership is transferred to this adapter. Initializing the adapter
+// will fail if the stream is already locked.
+//
+// If the adapter is dropped, or aborted while there are pending writes,
+// the pending writes will be rejected with the same exception as the abort.
+//
+// While WritableStream itself allows multiple writes to be in flight
+// at the same time, the WritableStreamSink interface does not, so
+// the adapter will ensure that only one write is in flight at a time.
+//
+// While the caller is expected to follow the WritableStreamSink contract
+// and keep the adapter alive until the write promises resolve, there
+// are some protections in place to avoid use-after-free if the caller
+// drops the adapter. There's nothing we can do if the caller drops the
+// buffer, however, so that is still a hard requirement.
+// TODO(safety): This can be made safer by having write take a kj::Array
+// as input instead of a kj::ArrayPtr but that's a larger refactor.
+//
+//     ┌───────────────────────────────────────────┐
+//     │         KJ Native Code                    │
+//     │                                           │
+//     │  • write(buffer)                          │
+//     │  • write(pieces[])                        │
+//     │  • end()                                  │
+//     │  • abort(reason)                          │
+//     │  • tryPumpFrom(source, end)               │
+//     └───────────────────────────────────────────┘
+//                            │
+//                            ▼
```

Comment:
```suggestion
//                            🭭
```

Reply:
Yeah, these showed up as replacement boxes in my mobile github client. I think I'll keep the current ones.